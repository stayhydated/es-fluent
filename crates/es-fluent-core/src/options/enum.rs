use std::str::FromStr as _;

use crate::strategy::DisplayStrategy;
use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta, FromVariant};
use getset::Getters;
use strum::IntoEnumIterator as _;

/// Options for an enum field.
#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct EnumFieldOpts {
    /// The identifier of the field.
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    /// The type of the field.
    #[getset(get = "pub")]
    ty: syn::Type,
    /// Whether to skip this field.
    #[darling(default)]
    skip: Option<bool>,
    /// Whether this field is a choice.
    #[darling(default)]
    choice: Option<bool>,
}

impl EnumFieldOpts {
    /// Returns `true` if the field should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns `true` if the field is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
}

/// Options for an enum variant.
#[derive(Clone, Debug, FromVariant, Getters)]
#[darling(attributes(fluent))]
pub struct VariantOpts {
    /// The identifier of the variant.
    #[getset(get = "pub")]
    ident: syn::Ident,
    fields: darling::ast::Fields<EnumFieldOpts>,
    /// Whether to skip this variant.
    #[darling(default)]
    skip: Option<bool>,
}

impl VariantOpts {
    /// Returns `true` if the variant should be skipped.
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }
    /// Returns the style of the variant's fields.
    pub fn style(&self) -> darling::ast::Style {
        self.fields.style
    }
    /// Returns the fields of the variant that are not skipped.
    pub fn fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields
            .iter()
            .filter(|field| !field.is_skipped())
            .collect()
    }
    /// Returns all fields of the variant.
    pub fn all_fields(&self) -> Vec<&EnumFieldOpts> {
        self.fields.iter().collect()
    }
}

/// Options for an enum.
#[derive(Clone, FromDeriveInput, Getters, Debug)]
#[darling(supports(enum_unit, enum_named, enum_tuple), attributes(fluent))]
#[getset(get = "pub")]
pub struct EnumOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<VariantOpts, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumFluentAttributeArgs,
}

impl EnumOpts {
    /// Returns the variants of the enum.
    pub fn variants(&self) -> Vec<&VariantOpts> {
        match &self.data {
            darling::ast::Data::Enum(variants) => variants.iter().collect(),
            _ => unreachable!("Unexpected data type for enum"),
        }
    }
}

/// Attribute arguments for an enum.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct EnumFluentAttributeArgs {
    display: Option<String>,
    #[darling(default)]
    choice: Option<bool>,
    #[darling(default)]
    this: Option<bool>,
}

impl EnumFluentAttributeArgs {
    /// Returns the display strategy for the enum.
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::FluentDisplay
        }
    }
    /// Returns `true` if the enum is a choice.
    pub fn is_choice(&self) -> bool {
        self.choice.unwrap_or(false)
    }
    /// Returns `true` if the enum should be passed as `this`.
    pub fn is_this(&self) -> bool {
        self.this.unwrap_or(false)
    }
}

/// Options for an enum that can be used as a choice.
#[derive(Clone, Debug, FromDeriveInput, Getters)]
#[darling(supports(enum_unit), attributes(fluent_choice))]
#[getset(get = "pub")]
pub struct EnumChoiceOpts {
    /// The identifier of the enum.
    ident: syn::Ident,
    /// The generics of the enum.
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, darling::util::Ignored>,
    #[darling(flatten)]
    attr_args: EnumChoiceAttributeArgs,
}

/// Attribute arguments for an enum that can be used as a choice.
#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
#[getset(get = "pub")]
pub struct EnumChoiceAttributeArgs {
    #[darling(default)]
    serialize_all: Option<String>,
}
