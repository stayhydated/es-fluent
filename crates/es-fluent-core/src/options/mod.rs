pub mod r#enum;
pub mod r#struct;

const DEFAULT_FL_PATH: &str = "crate::fl";
