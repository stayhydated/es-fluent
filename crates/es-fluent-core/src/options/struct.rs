use std::str::FromStr as _;

use bon::Builder;
use darling::{FromDeriveInput, FromField, FromMeta};
use getset::Getters;
use quote::format_ident;
use strum::IntoEnumIterator as _;

use crate::strategy::DisplayStrategy;

#[derive(Clone, Debug, FromField, Getters)]
#[darling(attributes(fluent))]
pub struct StructFieldOpts {
    #[getset(get = "pub")]
    ident: Option<syn::Ident>,
    #[getset(get = "pub")]
    ty: syn::Type,
    #[darling(default)]
    skip: Option<bool>,
    #[darling(default)]
    default: Option<bool>,
}

impl StructFieldOpts {
    pub fn is_skipped(&self) -> bool {
        self.skip.unwrap_or(false)
    }

    pub fn is_default(&self) -> bool {
        self.default.unwrap_or(false)
    }
}

#[derive(Clone, FromDeriveInput, Getters)]
#[darling(supports(struct_named), attributes(fluent))]
#[getset(get = "pub")]
pub struct StructOpts {
    ident: syn::Ident,
    #[darling(default)]
    derive: darling::util::PathList,
    generics: syn::Generics,
    data: darling::ast::Data<darling::util::Ignored, StructFieldOpts>,
    #[darling(flatten)]
    attr_args: StructFluentAttributeArgs,
}

impl StructOpts {
    const FTL_ENUM_IDENT: &str = "Ftl";

    pub fn ftl_enum_ident(&self) -> syn::Ident {
        format_ident!("{}{}", &self.ident, Self::FTL_ENUM_IDENT)
    }

    pub fn keyyed_idents(&self) -> Vec<syn::Ident> {
        self.attr_args
            .clone()
            .keys
            .map_or_else(Vec::new, |keys| {
                keys.into_iter()
                    .map(|key| format_ident!("{}", key.value()))
                    .collect()
            })
            .into_iter()
            .map(|key| format_ident!("{}{}{}", &self.ident, key, Self::FTL_ENUM_IDENT))
            .collect()
    }

    pub fn fields(&self) -> Vec<&StructFieldOpts> {
        match &self.data {
            darling::ast::Data::Struct(fields) => fields
                .fields
                .iter()
                .filter(|field| !field.is_skipped())
                .collect(),
            _ => vec![],
        }
    }
}

#[derive(Builder, Clone, Debug, Default, FromMeta, Getters)]
pub struct StructFluentAttributeArgs {
    #[darling(default)]
    fl: Option<syn::Path>, // this is probably useless, force user to provide a fl! macro?
    display: Option<String>,
    #[darling(default)]
    keys: Option<Vec<syn::LitStr>>,
}
impl StructFluentAttributeArgs {
    pub fn fl(&self) -> syn::Path {
        let default_fl_path: syn::Path = syn::parse_str(super::DEFAULT_FL_PATH).unwrap();
        self.fl.clone().unwrap_or(default_fl_path)
    }
    pub fn display(&self) -> DisplayStrategy {
        if let Some(mode_str) = self.display.as_deref() {
            DisplayStrategy::from_str(mode_str).unwrap_or_else(|_| {
                let possible_values: Vec<&'static str> = DisplayStrategy::iter()
                    .map(|variant| variant.into())
                    .collect();

                panic!(
                    "Unexpected mode: '{}', expected one of {:?}",
                    mode_str, possible_values
                );
            })
        } else {
            DisplayStrategy::StdDisplay
        }
    }
}
