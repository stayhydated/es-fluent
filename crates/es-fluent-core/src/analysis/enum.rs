use crate::meta::TypeKind;
use crate::namer;
use crate::options::r#enum::EnumOpts;
use crate::registry::{FtlTypeInfo, FtlVariant};

pub fn analyze_enum(opts: &EnumOpts, type_infos: &mut Vec<FtlTypeInfo>) {
    let target_ident = opts.ident();
    let target_name = target_ident.to_string();
    let opts_variants = opts.variants();
    let is_this = opts.attr_args().is_this();

    let mut unit_variants: Vec<FtlVariant> = opts_variants
        .iter()
        .filter_map(|variant_opt| {
            if variant_opt.is_skipped() {
                return None;
            }
            if matches!(variant_opt.style(), darling::ast::Style::Unit) {
                let name_str = variant_opt.ident().to_string();
                let ftl_key = namer::FluentKey::new(target_ident, &name_str);
                Some(
                    FtlVariant::builder()
                        .name(name_str)
                        .ftl_key(ftl_key)
                        .build(),
                )
            } else {
                None
            }
        })
        .collect();

    if is_this {
        let this_ftl_key = namer::FluentKey::new(target_ident, "");
        let this_variant = FtlVariant::builder()
            .name(target_name.clone())
            .ftl_key(this_ftl_key)
            .build();
        unit_variants.insert(0, this_variant);
    }

    if !unit_variants.is_empty() {
        log::debug!(
            "Generating FtlTypeInfo ({}) for '{}' (keys based on '{}') during {}",
            TypeKind::Enum,
            target_name,
            target_ident,
            "enum analysis (thiserror mode, unit variants)"
        );
        type_infos.push(
            FtlTypeInfo::builder()
                .type_kind(TypeKind::Enum)
                .type_name(target_name.clone())
                .variants(unit_variants)
                .build(),
        );
    }

    let mut struct_variants: Vec<FtlVariant> = opts_variants
        .iter()
        .filter_map(|variant_opt| {
            if variant_opt.is_skipped() {
                return None;
            }
            if matches!(variant_opt.style(), darling::ast::Style::Unit) {
                return None;
            }

            let name_str = variant_opt.ident().to_string();
            let arguments = match variant_opt.style() {
                darling::ast::Style::Struct => Some(
                    variant_opt
                        .fields()
                        .iter()
                        .filter_map(|f| f.ident().as_ref().map(|id| id.to_string()))
                        .collect(),
                ),
                darling::ast::Style::Tuple => Some(
                    variant_opt
                        .all_fields()
                        .iter()
                        .enumerate()
                        .filter_map(|(index, field)| {
                            if !field.is_skipped() {
                                Some(namer::UnnamedItem::from(index).to_string())
                            } else {
                                None
                            }
                        })
                        .collect(),
                ),
                darling::ast::Style::Unit => {
                    unreachable!("Unit variants should have been filtered out")
                },
            };

            let ftl_key = namer::FluentKey::new(target_ident, &name_str);
            Some(
                FtlVariant::builder()
                    .name(name_str)
                    .ftl_key(ftl_key)
                    .args(arguments.unwrap_or_default())
                    .build(),
            )
        })
        .collect();

    if is_this {
        let this_ftl_key = namer::FluentKey::new(target_ident, "");
        let this_variant = FtlVariant::builder()
            .name(target_name.clone())
            .ftl_key(this_ftl_key)
            .build();
        struct_variants.insert(0, this_variant);
    }

    if !struct_variants.is_empty() {
        log::debug!(
            "Generating FtlTypeInfo ({}) for '{}' (keys based on '{}') during {}",
            TypeKind::Struct,
            target_name,
            target_ident,
            "enum analysis (struct/tuple variants)"
        );
        type_infos.push(
            FtlTypeInfo::builder()
                .type_kind(TypeKind::Struct)
                .type_name(target_name)
                .variants(struct_variants)
                .build(),
        );
    }
}
