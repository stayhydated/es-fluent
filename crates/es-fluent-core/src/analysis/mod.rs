//! This module provides functions for analyzing Rust source code and extracting
//! information about types that can be translated with `es-fluent`.

mod r#enum;
mod r#struct;

use crate::options::r#enum::EnumOpts;
use crate::options::r#struct::StructOpts;
use crate::registry::FtlTypeInfo;

/// Analyzes a struct and returns a list of `FtlTypeInfo` objects.
pub fn analyze_struct(opts: &StructOpts) -> Vec<FtlTypeInfo> {
    let mut type_infos = Vec::new();

    r#struct::analyze_struct(opts, &mut type_infos);

    type_infos
}

/// Analyzes an enum and returns a list of `FtlTypeInfo` objects.
pub fn analyze_enum(opts: &EnumOpts) -> Vec<FtlTypeInfo> {
    let mut type_infos = Vec::new();

    r#enum::analyze_enum(opts, &mut type_infos);

    type_infos
}
