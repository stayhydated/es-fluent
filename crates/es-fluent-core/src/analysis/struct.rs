use crate::meta::TypeKind;
use crate::namer;
use crate::options::r#struct::StructOpts;
use crate::registry::{FtlTypeInfo, FtlVariant};

pub fn analyze_struct(opts: &StructOpts, type_infos: &mut Vec<FtlTypeInfo>) {
    let target_ident = opts.ident();
    let keyyed_idents = opts.keyyed_idents();
    let is_this = opts.attr_args().is_this();

    let field_names: Vec<String> = opts
        .fields()
        .iter()
        .filter_map(|field_opt| {
            field_opt
                .ident()
                .as_ref()
                .map(|field_ident| field_ident.to_string())
        })
        .collect();

    if field_names.is_empty() {
        return;
    }

    if keyyed_idents.is_empty() {
        let ftl_enum_ident = opts.ftl_enum_ident();

        let mut variants: Vec<FtlVariant> = field_names
            .iter()
            .map(|name_str| {
                let ftl_key = namer::FluentKey::new(&ftl_enum_ident, name_str);
                FtlVariant::builder()
                    .name(name_str.clone())
                    .ftl_key(ftl_key)
                    .build()
            })
            .collect();

        if is_this {
            let this_ftl_key = namer::FluentKey::new(&ftl_enum_ident, "");
            let this_variant = FtlVariant::builder()
                .name(ftl_enum_ident.to_string())
                .ftl_key(this_ftl_key)
                .build();
            variants.insert(0, this_variant);
        }

        log::debug!(
            "Generating FtlTypeInfo ({}) for '{}' (keys based on '{}') during {}",
            TypeKind::Enum,
            ftl_enum_ident,
            target_ident,
            "struct analysis (FTL enum generation)"
        );
        type_infos.push(
            FtlTypeInfo::builder()
                .type_kind(TypeKind::Enum)
                .type_name(ftl_enum_ident.to_string())
                .variants(variants)
                .build(),
        );
    } else {
        for keyyed_ident in keyyed_idents {
            let mut variants: Vec<FtlVariant> = field_names
                .iter()
                .map(|name_str| {
                    let ftl_key = namer::FluentKey::new(&keyyed_ident, name_str);
                    FtlVariant::builder()
                        .name(name_str.clone())
                        .ftl_key(ftl_key)
                        .build()
                })
                .collect();

            if is_this {
                let this_ftl_key = namer::FluentKey::new(&keyyed_ident, "");
                let this_variant = FtlVariant::builder()
                    .name(keyyed_ident.to_string())
                    .ftl_key(this_ftl_key)
                    .build();
                variants.insert(0, this_variant);
            }

            log::debug!(
                "Generating FtlTypeInfo ({}) for '{}' (keys based on '{}') during {}",
                TypeKind::Enum,
                keyyed_ident,
                target_ident,
                "struct analysis (FTL enum generation with keys)"
            );
            type_infos.push(
                FtlTypeInfo::builder()
                    .type_kind(TypeKind::Enum)
                    .type_name(keyyed_ident.to_string())
                    .variants(variants)
                    .build(),
            );
        }
    }
}
