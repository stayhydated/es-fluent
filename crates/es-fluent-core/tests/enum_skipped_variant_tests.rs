use darling::FromDeriveInput;
use es_fluent_core::analysis;
use es_fluent_core::options::r#enum::EnumOpts;
use syn::{DeriveInput, parse_quote};

#[test]
fn enum_analysis_with_skipped_variants_without_this() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent]
        enum Mixed {
            // Unit variants
            #[fluent(skip)]
            UnitSkipped,
            UnitKept,

            // Struct variants
            #[fluent(skip)]
            DataSkipped { x: i32, y: i32 },
            DataKept {
                x: i32,
                #[fluent(skip)]
                secret: String,
                z: bool,
            },

            // Tuple variants
            #[fluent(skip)]
            TupSkipped(#[fluent(skip)] i64, String),
            TupKept(#[fluent(skip)] i64, String, u8),
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    let infos = analysis::analyze_enum(&opts);

    // Skipped variants are filtered during analysis.
    // This snapshot documents that only non-skipped variants are present.
    insta::assert_ron_snapshot!("enum_analysis_with_skipped_variants_without_this", &infos);
}

#[test]
fn enum_analysis_with_skipped_variants_with_this() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent(this)]
        enum Mixed {
            // Unit variants
            #[fluent(skip)]
            UnitSkipped,
            UnitKept,

            // Struct variants
            #[fluent(skip)]
            DataSkipped { x: i32, y: i32 },
            DataKept {
                x: i32,
                #[fluent(skip)]
                secret: String,
                z: bool,
            },

            // Tuple variants
            #[fluent(skip)]
            TupSkipped(#[fluent(skip)] i64, String),
            TupKept(#[fluent(skip)] i64, String, u8),
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    let infos = analysis::analyze_enum(&opts);

    // Skipped variants are filtered during analysis.
    // With #[fluent(this)], the "this" variant appears first in both unit and struct/tuple groups, but only non-skipped variants are included.
    insta::assert_ron_snapshot!("enum_analysis_with_skipped_variants_with_this", &infos);
}

#[test]
fn enum_analysis_all_variants_skipped_yields_empty() {
    let input: DeriveInput = parse_quote! {
        #[derive(EsFluent)]
        #[fluent]
        enum AllSkipped {
            #[fluent(skip)]
            UnitA,
            #[fluent(skip)]
            Data { x: i32 },
            #[fluent(skip)]
            Tup(i32),
        }
    };

    let opts = EnumOpts::from_derive_input(&input).expect("EnumOpts should parse");
    let infos = analysis::analyze_enum(&opts);

    // When all variants are marked #[fluent(skip)], no FtlTypeInfo is produced.
    // This snapshot should be an empty list.
    insta::assert_ron_snapshot!("enum_analysis_all_variants_skipped_yields_empty", &infos);
}
