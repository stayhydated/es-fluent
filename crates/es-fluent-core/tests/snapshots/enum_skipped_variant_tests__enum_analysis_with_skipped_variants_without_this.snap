---
source: crates/es-fluent-core/tests/enum_skipped_variant_tests.rs
expression: "&infos"
---
[
  FtlTypeInfo(
    type_kind: Enum,
    type_name: "Mixed",
    variants: [
      FtlVariant(
        name: "UnitKept",
        ftl_key: FluentKey("mixed-UnitKept"),
        arguments: None,
      ),
    ],
  ),
  FtlTypeInfo(
    type_kind: Struct,
    type_name: "Mixed",
    variants: [
      FtlVariant(
        name: "DataKept",
        ftl_key: FluentKey("mixed-DataKept"),
        arguments: Some([
          "x",
          "z",
        ]),
      ),
      FtlVariant(
        name: "TupKept",
        ftl_key: FluentKey("mixed-TupKept"),
        arguments: Some([
          "f1",
          "f2",
        ]),
      ),
    ],
  ),
]
